src/transport_ucx.h

```diff
@@ -0,0 +1,846 @@
+                                  target, len, (uint64_t) remote_addr, rkey,
+                                  &shmem_transport_ucx_cb_nop);
+
+    ucs_status_t status = shmem_transport_ucx_complete_op(pstatus);
+    UCX_CHECK_STATUS(status);
+}
+
+static inline
+void
+shmem_transport_atomic_set(shmem_transport_ctx_t* ctx, void *target, const void *source, size_t len,
+                             int pe, shm_internal_datatype_t datatype)
+{
+    uint8_t *remote_addr;
+    ucp_rkey_h rkey;
+    ucs_status_ptr_t pstatus;
+    uint64_t value;
+
+    /* XXX: Set is implemented as swap, so dest is thrown away. Allocate dest
+     * as a static rather than on the stack to avoid needing to block on
+     * completion before returning. */
+    static uint64_t dest;
+
+    shmem_transport_ucx_get_mr(target, pe, &remote_addr, &rkey);
+
+    switch (len) {
+        case 1:
+            value = (uint64_t) *(uint8_t*)source;
+            break;
+        case 2:
+            value = (uint64_t) *(uint16_t*)source;
+            break;
+        case 4:
+            value = (uint64_t) *(uint32_t*)source;
+            break;
+        case 8:
+            value = (uint64_t) *(uint64_t*)source;
+            break;
+        default:
+            RAISE_ERROR_MSG("Unsupported datatype len=%zu\n", len);
+    }
+
+    pstatus = ucp_atomic_fetch_nb(shmem_transport_peers[pe].ep, UCP_ATOMIC_FETCH_OP_SWAP, value,
+                                  &dest, len, (uint64_t) remote_addr, rkey,
+                                  &shmem_transport_ucx_cb_nop);
+
+    ucs_status_t status = shmem_transport_ucx_release_op(pstatus);
+    UCX_CHECK_STATUS_INPROGRESS(status);
+}
+
+static inline
+void
+shmem_transport_mswap(shmem_transport_ctx_t* ctx, void *target, const void *source, void *dest,
+                      const void *mask, size_t len, int pe,
+                      shm_internal_datatype_t datatype)
+{
+    uint8_t *remote_addr;
+    ucp_rkey_h rkey;
+    int done = 0;
+
+    shmem_transport_ucx_get_mr(target, pe, &remote_addr, &rkey);
+
+    if (len != 4)
+        RAISE_ERROR_STR("Unsupported datatype");
+
+    /* XXX: Emulate MSWAP, since it is not exposed by UCX */
+    while (!done) {
+        uint32_t v;
+
+        shmem_transport_atomic_fetch(ctx, &v, target, len, pe, datatype);
+
+        uint32_t new = (v & ~*(uint32_t *)mask) | (*(uint32_t *)source & *(uint32_t *)mask);
+
+        shmem_transport_cswap(ctx, target, &new, dest, &v, len, pe, datatype);
+        if (*(uint32_t *)dest == v) done = 1;
+
+        /* Manual progress to avoid deadlock for application-level polling */
+        shmem_transport_probe();
+    }
+}
+
+static inline
+int shmem_transport_atomic_supported(shm_internal_op_t op, shm_internal_datatype_t datatype)
+{
+    /* Use software reductions, instead of atomic vector operation */
+    return 0;
+}
+
+static inline
+void shmem_transport_syncmem(void)
+{
+    return;
+}
```

Comment:
@jdinan Trying to understand, is it similar to libfabric where there is nothing similar to PtlAtomicSync or is it done in the hardware?

Reply:
I'm not aware of one. I'll ask the experts and let you know what I hear.

Reply:
Nope, nothing special that needs to be done here.